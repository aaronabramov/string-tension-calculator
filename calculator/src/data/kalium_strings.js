// @flow

import type {PhysicalString, Gauge} from '../types.js';

export const data: Array<PhysicalString> = [
  {gauge: 0.008, unitWeight: 0.00001424},
  {gauge: 0.0085, unitWeight: 0.000016075},
  {gauge: 0.009, unitWeight: 0.000018022},
  {gauge: 0.0095, unitWeight: 0.00002008},
  {gauge: 0.01, unitWeight: 0.000022252},
  {gauge: 0.0105, unitWeight: 0.000024531},
  {gauge: 0.011, unitWeight: 0.000026925},
  {gauge: 0.0115, unitWeight: 0.000029425},
  {gauge: 0.012, unitWeight: 0.000032039},
  {gauge: 0.0125, unitWeight: 0.000034765},
  {gauge: 0.013, unitWeight: 0.000037605},
  {gauge: 0.0135, unitWeight: 0.000040551},
  {gauge: 0.014, unitWeight: 0.000043607},
  {gauge: 0.015, unitWeight: 0.00005005},
  {gauge: 0.016, unitWeight: 0.000056961},
  {gauge: 0.017, unitWeight: 0.0000643},
  {gauge: 0.018, unitWeight: 0.000072088},
  {gauge: 0.019, unitWeight: 0.00008036},
  {gauge: 0.02, unitWeight: 0.000089031},
  {gauge: 0.021, unitWeight: 0.00009387286685, type: 'w'},
  {gauge: 0.021, unitWeight: 0.0000981546917942, type: 'p'},
  {gauge: 0.022, unitWeight: 0.000103862422, type: 'w'},
  {gauge: 0.022, unitWeight: 0.00010766580534, type: 'p'},
  {gauge: 0.023, unitWeight: 0.00011398499841, type: 'w'},
  {gauge: 0.023, unitWeight: 0.000117702448211, type: 'p'},
  {gauge: 0.024, unitWeight: 0.000124963},
  {gauge: 0.025, unitWeight: 0.000136054},
  {gauge: 0.026, unitWeight: 0.000144691},
  {gauge: 0.027, unitWeight: 0.000153146},
  {gauge: 0.028, unitWeight: 0.000161203},
  {gauge: 0.029, unitWeight: 0.000178551},
  {gauge: 0.031, unitWeight: 0.000198902},
  {gauge: 0.033, unitWeight: 0.000223217},
  {gauge: 0.035, unitWeight: 0.000249034},
  {gauge: 0.037, unitWeight: 0.000276237},
  {gauge: 0.039, unitWeight: 0.000304788},
  {gauge: 0.041, unitWeight: 0.000334965},
  {gauge: 0.043, unitWeight: 0.000366357},
  {gauge: 0.045, unitWeight: 0.000404956},
  {gauge: 0.047, unitWeight: 0.000447408},
  {gauge: 0.049, unitWeight: 0.000475438},
  {gauge: 0.051, unitWeight: 0.000512645},
  {gauge: 0.053, unitWeight: 0.000551898},
  {gauge: 0.055, unitWeight: 0.000584407},
  {gauge: 0.057, unitWeight: 0.000625704},
  {gauge: 0.059, unitWeight: 0.000679149},
  {gauge: 0.061, unitWeight: 0.000720293},
  {gauge: 0.063, unitWeight: 0.000765973},
  {gauge: 0.065, unitWeight: 0.000821116},
  {gauge: 0.067, unitWeight: 0.000870707},
  {gauge: 0.07, unitWeight: 0.000939851},
  {gauge: 0.073, unitWeight: 0.001021518},
  {gauge: 0.076, unitWeight: 0.001110192},
  {gauge: 0.079, unitWeight: 0.001188974},
  {gauge: 0.082, unitWeight: 0.001293598},
  {gauge: 0.086, unitWeight: 0.001416131},
  {gauge: 0.09, unitWeight: 0.001544107},
  {gauge: 0.094, unitWeight: 0.001677765},
  {gauge: 0.098, unitWeight: 0.001831487},
  {gauge: 0.102, unitWeight: 0.001986524},
  {gauge: 0.106, unitWeight: 0.002127413},
  {gauge: 0.11, unitWeight: 0.002286733},
  {gauge: 0.114, unitWeight: 0.002449286},
  {gauge: 0.118, unitWeight: 0.002616406},
  {gauge: 0.124, unitWeight: 0.002880915},
  {gauge: 0.13, unitWeight: 0.003154996},
  {gauge: 0.136, unitWeight: 0.003441822},
  {gauge: 0.142, unitWeight: 0.003741715},
  {gauge: 0.148, unitWeight: 0.004051505},
  {gauge: 0.154, unitWeight: 0.004375389},
  {gauge: 0.16, unitWeight: 0.004712133},
  {gauge: 0.168, unitWeight: 0.005204133},
  {gauge: 0.176, unitWeight: 0.005691179},
  {gauge: 0.184, unitWeight: 0.006219447},
  {gauge: 0.192, unitWeight: 0.006742017},
  {gauge: 0.2, unitWeight: 0.007409517},
  {gauge: 0.208, unitWeight: 0.007875038},
  {gauge: 0.216, unitWeight: 0.008477001},
  {gauge: 0.226, unitWeight: 0.009397395},
  {gauge: 0.236, unitWeight: 0.010222903},
  {gauge: 0.246, unitWeight: 0.011082025},
  {gauge: 0.256, unitWeight: 0.011969036},
  {gauge: 0.266, unitWeight: 0.012897577},
];

export const findByGauge = (gauge: Gauge): PhysicalString => {
  const string = data.find(string => string.gauge === gauge);
  if (!string) {
    throw new Error(`could not find a string with the gauge: %{gauge}`);
  }
  return string;
};

export const findNext = (string: PhysicalString): ?PhysicalString => {
  return data.find(next => next.unitWeight > string.unitWeight);
};

export const findPrevious = (string: PhysicalString): ?PhysicalString => {
  for (let i = data.length - 1; i >= 0; i--) {
    if (data[i] && data[i].unitWeight < string.unitWeight) {
      return data[i];
    }
  }
};

export const formatStringGauge = (string: PhysicalString) => {
  return `${string.gauge}${string.type || ''}`;
};
