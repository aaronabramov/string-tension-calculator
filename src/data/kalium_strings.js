// @flow

import type {PhysicalString, Gauge} from '../types.js';

export const data: Array<PhysicalString> = [
  [0.008, 0.00001424],
  [0.0085, 0.000016075],
  [0.009, 0.000018022],
  [0.0095, 0.00002008],
  [0.01, 0.000022252],
  [0.0105, 0.000024531],
  [0.011, 0.000026925],
  [0.0115, 0.000029425],
  [0.012, 0.000032039],
  [0.0125, 0.000034765],
  [0.013, 0.000037605],
  [0.0135, 0.000040551],
  [0.014, 0.000043607],
  [0.015, 0.00005005],
  [0.016, 0.000056961],
  [0.017, 0.0000643],
  [0.018, 0.000072088],
  [0.019, 0.00008036],
  [0.02, 0.000089031],
  [0.021, 0.000098155],
  [0.022, 0.000107666],
  [0.023, 0.000117702],
  [0.024, 0.000124963],
  [0.025, 0.000136054],
  [0.026, 0.000144691],
  [0.027, 0.000153146],
  [0.028, 0.000161203],
  [0.029, 0.000178551],
  [0.031, 0.000198902],
  [0.033, 0.000223217],
  [0.035, 0.000249034],
  [0.037, 0.000276237],
  [0.039, 0.000304788],
  [0.041, 0.000334965],
  [0.043, 0.000366357],
  [0.045, 0.000404956],
  [0.047, 0.000447408],
  [0.049, 0.000475438],
  [0.051, 0.000512645],
  [0.053, 0.000551898],
  [0.055, 0.000584407],
  [0.057, 0.000625704],
  [0.059, 0.000679149],
  [0.061, 0.000720293],
  [0.063, 0.000765973],
  [0.065, 0.000821116],
  [0.067, 0.000870707],
  [0.07, 0.000939851],
  [0.073, 0.001021518],
  [0.076, 0.001110192],
  [0.079, 0.001188974],
  [0.082, 0.001293598],
  [0.086, 0.001416131],
  [0.09, 0.001544107],
  [0.094, 0.001677765],
  [0.098, 0.001831487],
  [0.102, 0.001986524],
  [0.106, 0.002127413],
  [0.11, 0.002286733],
  [0.114, 0.002449286],
  [0.118, 0.002616406],
  [0.124, 0.002880915],
  [0.13, 0.003154996],
  [0.136, 0.003441822],
  [0.142, 0.003741715],
  [0.148, 0.004051505],
  [0.154, 0.004375389],
  [0.16, 0.004712133],
  [0.168, 0.005204133],
  [0.176, 0.005691179],
  [0.184, 0.006219447],
  [0.192, 0.006742017],
  [0.2, 0.007409517],
  [0.208, 0.007875038],
  [0.216, 0.008477001],
  [0.226, 0.009397395],
  [0.236, 0.010222903],
  [0.246, 0.011082025],
  [0.256, 0.011969036],
  [0.266, 0.012897577],
].map(s => ({gauge: s[0], unitWeight: s[1]}));

export const findByGauge = (gauge: Gauge): PhysicalString => {
  const string = data.find(string => string.gauge === gauge);
  if (!string) {
    throw new Error(`could not find a string with the gauge: %{gauge}`);
  }
  return string;
};

export const findNext = (gauge: Gauge): ?PhysicalString => {
  return data.find(string => string.gauge > gauge);
};

export const findPrevious = (gauge: Gauge): ?PhysicalString => {
  for (let i = data.length - 1; i >= 0; i--) {
    if (data[i] && data[i].gauge < gauge) {
      return data[i];
    }
  }
};
