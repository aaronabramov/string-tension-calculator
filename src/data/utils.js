// @flow

export const roundTo = (number: number, digits: number) => {
  return Math.floor(number * Math.pow(10, digits)) / Math.pow(10, digits);
};
