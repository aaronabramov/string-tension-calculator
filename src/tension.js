// @flow

import Color from 'color';

export type Instrument = 'guitar' | 'bass';

const TENSION_CONST = 386.4;

export const getTension = ({unitWeight, scale, freq}: {unitWeight: number, scale: number, freq: number}): number => {
  return (unitWeight * Math.pow(2 * scale * freq, 2)) / TENSION_CONST;
};

export class Note {
  _midi: number;
  constructor(midi: number) {
    this._midi = midi;
  }

  static fromMidi(midi: number) {
    return new Note(midi);
  }

  static fromScientific(scientific: string) {
    const midi = scientificToMidi(scientific);
    return Note.fromMidi(midi);
  }

  freq() {
    return midiToFreq(this._midi);
  }

  scientific() {
    return midiToScientific(this._midi);
  }

  midi() {
    return this._midi;
  }
}

const SCIENTIFIC = {
  c: 0,
  d: 2,
  e: 4,
  f: 5,
  g: 7,
  a: 9,
  b: 11,
};

const MIDI = ['c', 'c#', 'd', 'd#', 'e', 'f', 'f#', 'g', 'g#', 'a', 'a#', 'b'];

// 0 midi is A-5
const FIRST_OCTAVE = -1;

const SCIENTIFIC_REGEXP = /^([ABCDEFG])(#?)(-?\d*)$/i;
export const scientificToMidi = (scientific: string): number => {
  const match = scientific.match(SCIENTIFIC_REGEXP);
  if (!match) {
    throw new Error(`Expected the note to be in scientific notation ${String(SCIENTIFIC_REGEXP)}, got ${scientific}`);
  }

  const letter = match[1].toLowerCase();
  const sharp = !!match[2];
  const octave = parseInt(match[3], 10);
  if (octave < -1) {
    throw new Error('-1 is the lowest octave available');
  }
  if (['b', 'e'].includes(letter) && sharp === true) {
    throw new Error(`B and E don't have sharps. got: ${scientific}`);
  }

  let baseMidi = SCIENTIFIC[letter];
  if (sharp) {
    baseMidi += 1;
  }
  return baseMidi + (octave - FIRST_OCTAVE) * 12;
};

export const midiToScientific = (midi: number) => {
  const octave = Math.floor(midi / 12);
  const baseMidi = midi % 12;
  const baseNote = MIDI[baseMidi];
  if (!baseNote) {
    throw new Error(
      `can't find the base note baseMidi: ${baseMidi}, base notes: ${JSON.stringify(MIDI)}, midi: ${midi}`,
    );
  }
  return `${baseNote.toUpperCase()}${octave + FIRST_OCTAVE}`;
};

const C0_FREQ = 8.1757989156;
const NOTE_TO_FREQ_CONST = Math.pow(2, 1 / 12);

export const midiToFreq = (midi: number) => {
  return C0_FREQ * Math.pow(NOTE_TO_FREQ_CONST, midi);
};

// See description of https://www.youtube.com/watch?v=ZxiaiA8ejqs
const TIGHTNESS_DATA = {
  guitar: [
    {
      lowerScale: 25,
      higherScale: 27,
      mediumTension: 18,
      margin: 2,
    },
    {
      lowerScale: 27,
      higherScale: 29,
      mediumTension: 23,
      margin: 3,
    },
    {
      lowerScale: 29,
      higherScale: 31,
      mediumTension: 29,
      margin: 4,
    },
  ],
  bass: [
    {
      lowerScale: 30,
      higherScale: 34,
      mediumTension: 29,
      margin: 4,
    },
    {
      lowerScale: 34,
      higherScale: 36,
      mediumTension: 40,
      margin: 8,
    },
    {
      lowerScale: 36,
      higherScale: 38,
      mediumTension: 41,
      margin: 8,
    },
    {
      lowerScale: 39,
      higherScale: 42,
      mediumTension: 42,
      margin: 8,
    },
  ],
};

type HowTight = {|direction: 'HEAVY' | 'LIGHT', howMuch: number|} | {|direction: 'UNKNOWN'|};

export const howTight = ({
  tension,
  scale,
  instrument,
}: {
  tension: number,
  scale: number,
  instrument: Instrument,
}): HowTight => {
  const data = TIGHTNESS_DATA[instrument].find(
    ({lowerScale, higherScale}) => scale >= lowerScale && scale <= higherScale,
  );
  if (!data) {
    return {direction: 'UNKNOWN'};
  }
  const direction = tension >= data.mediumTension ? 'HEAVY' : 'LIGHT';
  const howMuch = Math.abs((tension - data.mediumTension) / (data.margin / 100));
  return {direction, howMuch};
};

export const getTightnessColor = ({
  tension,
  scale,
  instrument,
}: {
  tension: number,
  scale: number,
  instrument: Instrument,
}) => {
  const result = howTight({tension, scale, instrument});
  if (result.direction === 'UNKNOWN') {
    return Color.rgb(200, 200, 200);
  }
  const {direction, howMuch} = result;
  const color = direction === 'HEAVY' ? Color.rgb(255, 0, 0) : Color.rgb(255, 255, 0);
  const intensity = 1 - Math.min(howMuch / 500, 1); // 0 to 1
  return color.lighten(intensity).string();
};
