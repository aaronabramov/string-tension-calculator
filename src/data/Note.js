// @flow

import {midiToFreq, scientificToMidi, midiToScientific} from './tension.js';

export class Note {
  _midi: number;
  constructor(midi: number) {
    this._midi = midi;
  }

  static fromMidi(midi: number) {
    return new Note(midi);
  }

  static fromScientific(scientific: string) {
    const midi = scientificToMidi(scientific);
    return Note.fromMidi(midi);
  }

  freq(): number {
    return midiToFreq(this._midi);
  }

  scientific(): string {
    return midiToScientific(this._midi);
  }

  midi() {
    return this._midi;
  }

  // return a new note that is `steps` steps away from the
  // original note
  interval(steps: number) {
    return new Note(this._midi + steps);
  }
}
