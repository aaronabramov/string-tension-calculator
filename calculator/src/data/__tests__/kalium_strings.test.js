// @flow

import {findByGauge, findNext, findPrevious, data} from '../kalium_strings';
import {getTension} from '../tension.js';
import {Note} from '../Note.js';
import {roundTo} from '../utils.js';
import leftPad from 'left-pad';

(test: any).each([0.059, 0.11, 0.114])('finds %p string', gauge => {
  expect(findByGauge(gauge)).toHaveProperty('gauge', gauge);
  expect(findByGauge(gauge)).toHaveProperty('unitWeight');
});

(test: any).each([0.0333, 0.33, 0, 1])('doesnt find %p string', gauge => {
  expect(() => findByGauge(gauge)).toThrow(/gauge/);
});

test('string search', () => {
  let result = [];
  let gauge = 0.005;
  while (gauge < 0.28) {
    const prev = findPrevious(gauge);
    const next = findNext(gauge);

    const cg = leftPad(roundTo(gauge, 4), 7);
    const pg = prev ? leftPad(roundTo(prev.gauge, 4), 7) : leftPad('NONE', 7);
    const ng = next ? leftPad(roundTo(next.gauge, 4), 7) : leftPad('NONE', 7);
    result.push(`gauge: ${cg} | prev: ${pg} | next: ${ng}`);

    gauge += 0.001;
  }
  expect(result).toMatchSnapshot();
});

test('snapshot all tensions for a freq', () => {
  const note = Note.fromScientific('A3');
  expect(
    data.map(({gauge, unitWeight}) => {
      const tension = getTension({unitWeight, scale: 25.5, freq: note.freq()});
      return `gauge: ${leftPad(gauge, 10)} | tension: ${leftPad(roundTo(tension, 2), 10)}`;
    }),
  ).toMatchSnapshot();
});
