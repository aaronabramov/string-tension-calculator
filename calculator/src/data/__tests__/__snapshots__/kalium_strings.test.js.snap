// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot all tensions for a freq 1`] = `
Array [
  "gauge:      0.008 | tension:       4.63",
  "gauge:     0.0085 | tension:       5.23",
  "gauge:      0.009 | tension:       5.87",
  "gauge:     0.0095 | tension:       6.54",
  "gauge:       0.01 | tension:       7.24",
  "gauge:     0.0105 | tension:       7.99",
  "gauge:      0.011 | tension:       8.77",
  "gauge:     0.0115 | tension:       9.58",
  "gauge:      0.012 | tension:      10.43",
  "gauge:     0.0125 | tension:      11.32",
  "gauge:      0.013 | tension:      12.25",
  "gauge:     0.0135 | tension:      13.21",
  "gauge:      0.014 | tension:       14.2",
  "gauge:      0.015 | tension:       16.3",
  "gauge:      0.016 | tension:      18.55",
  "gauge:      0.017 | tension:      20.94",
  "gauge:      0.018 | tension:      23.48",
  "gauge:      0.019 | tension:      26.18",
  "gauge:       0.02 | tension:         29",
  "gauge:      0.021 | tension:      30.58",
  "gauge:      0.021 | tension:      31.97",
  "gauge:      0.022 | tension:      33.83",
  "gauge:      0.022 | tension:      35.07",
  "gauge:      0.023 | tension:      37.13",
  "gauge:      0.023 | tension:      38.34",
  "gauge:      0.024 | tension:      40.71",
  "gauge:      0.025 | tension:      44.32",
  "gauge:      0.026 | tension:      47.14",
  "gauge:      0.027 | tension:      49.89",
  "gauge:      0.028 | tension:      52.51",
  "gauge:      0.029 | tension:      58.17",
  "gauge:      0.031 | tension:       64.8",
  "gauge:      0.033 | tension:      72.72",
  "gauge:      0.035 | tension:      81.13",
  "gauge:      0.037 | tension:      89.99",
  "gauge:      0.039 | tension:      99.29",
  "gauge:      0.041 | tension:     109.13",
  "gauge:      0.043 | tension:     119.35",
  "gauge:      0.045 | tension:     131.93",
  "gauge:      0.047 | tension:     145.76",
  "gauge:      0.049 | tension:     154.89",
  "gauge:      0.051 | tension:     167.01",
  "gauge:      0.053 | tension:      179.8",
  "gauge:      0.055 | tension:     190.39",
  "gauge:      0.057 | tension:     203.85",
  "gauge:      0.059 | tension:     221.26",
  "gauge:      0.061 | tension:     234.67",
  "gauge:      0.063 | tension:     249.55",
  "gauge:      0.065 | tension:     267.51",
  "gauge:      0.067 | tension:     283.67",
  "gauge:       0.07 | tension:      306.2",
  "gauge:      0.073 | tension:      332.8",
  "gauge:      0.076 | tension:     361.69",
  "gauge:      0.079 | tension:     387.36",
  "gauge:      0.082 | tension:     421.45",
  "gauge:      0.086 | tension:     461.37",
  "gauge:       0.09 | tension:     503.06",
  "gauge:      0.094 | tension:     546.61",
  "gauge:      0.098 | tension:     596.69",
  "gauge:      0.102 | tension:      647.2",
  "gauge:      0.106 | tension:      693.1",
  "gauge:       0.11 | tension:     745.01",
  "gauge:      0.114 | tension:     797.97",
  "gauge:      0.118 | tension:     852.42",
  "gauge:      0.124 | tension:     938.59",
  "gauge:       0.13 | tension:    1027.89",
  "gauge:      0.136 | tension:    1121.33",
  "gauge:      0.142 | tension:    1219.04",
  "gauge:      0.148 | tension:    1319.97",
  "gauge:      0.154 | tension:    1425.49",
  "gauge:       0.16 | tension:     1535.2",
  "gauge:      0.168 | tension:    1695.49",
  "gauge:      0.176 | tension:    1854.17",
  "gauge:      0.184 | tension:    2026.28",
  "gauge:      0.192 | tension:    2196.53",
  "gauge:        0.2 | tension:       2414",
  "gauge:      0.208 | tension:    2565.67",
  "gauge:      0.216 | tension:    2761.79",
  "gauge:      0.226 | tension:    3061.65",
  "gauge:      0.236 | tension:     3330.6",
  "gauge:      0.246 | tension:     3610.5",
  "gauge:      0.256 | tension:    3899.48",
  "gauge:      0.266 | tension:       4202",
]
`;

exports[`string search 1`] = `
Array [
  "string:   0.008 | prev:     NONE | next:   0.0085",
  "string:  0.0085 | prev:    0.008 | next:    0.009",
  "string:   0.009 | prev:   0.0085 | next:   0.0095",
  "string:  0.0095 | prev:    0.009 | next:     0.01",
  "string:    0.01 | prev:   0.0095 | next:   0.0105",
  "string:  0.0105 | prev:     0.01 | next:    0.011",
  "string:   0.011 | prev:   0.0105 | next:   0.0115",
  "string:  0.0115 | prev:    0.011 | next:    0.012",
  "string:   0.012 | prev:   0.0115 | next:   0.0125",
  "string:  0.0125 | prev:    0.012 | next:    0.013",
  "string:   0.013 | prev:   0.0125 | next:   0.0135",
  "string:  0.0135 | prev:    0.013 | next:    0.014",
  "string:   0.014 | prev:   0.0135 | next:    0.015",
  "string:   0.015 | prev:    0.014 | next:    0.016",
  "string:   0.016 | prev:    0.015 | next:    0.017",
  "string:   0.017 | prev:    0.016 | next:    0.018",
  "string:   0.018 | prev:    0.017 | next:    0.019",
  "string:   0.019 | prev:    0.018 | next:     0.02",
  "string:    0.02 | prev:    0.019 | next:   0.021w",
  "string:  0.021w | prev:     0.02 | next:   0.021p",
  "string:  0.021p | prev:   0.021w | next:   0.022w",
  "string:  0.022w | prev:   0.021p | next:   0.022p",
  "string:  0.022p | prev:   0.022w | next:   0.023w",
  "string:  0.023w | prev:   0.022p | next:   0.023p",
  "string:  0.023p | prev:   0.023w | next:    0.024",
  "string:   0.024 | prev:   0.023p | next:    0.025",
  "string:   0.025 | prev:    0.024 | next:    0.026",
  "string:   0.026 | prev:    0.025 | next:    0.027",
  "string:   0.027 | prev:    0.026 | next:    0.028",
  "string:   0.028 | prev:    0.027 | next:    0.029",
  "string:   0.029 | prev:    0.028 | next:    0.031",
  "string:   0.031 | prev:    0.029 | next:    0.033",
  "string:   0.033 | prev:    0.031 | next:    0.035",
  "string:   0.035 | prev:    0.033 | next:    0.037",
  "string:   0.037 | prev:    0.035 | next:    0.039",
  "string:   0.039 | prev:    0.037 | next:    0.041",
  "string:   0.041 | prev:    0.039 | next:    0.043",
  "string:   0.043 | prev:    0.041 | next:    0.045",
  "string:   0.045 | prev:    0.043 | next:    0.047",
  "string:   0.047 | prev:    0.045 | next:    0.049",
  "string:   0.049 | prev:    0.047 | next:    0.051",
  "string:   0.051 | prev:    0.049 | next:    0.053",
  "string:   0.053 | prev:    0.051 | next:    0.055",
  "string:   0.055 | prev:    0.053 | next:    0.057",
  "string:   0.057 | prev:    0.055 | next:    0.059",
  "string:   0.059 | prev:    0.057 | next:    0.061",
  "string:   0.061 | prev:    0.059 | next:    0.063",
  "string:   0.063 | prev:    0.061 | next:    0.065",
  "string:   0.065 | prev:    0.063 | next:    0.067",
  "string:   0.067 | prev:    0.065 | next:     0.07",
  "string:    0.07 | prev:    0.067 | next:    0.073",
  "string:   0.073 | prev:     0.07 | next:    0.076",
  "string:   0.076 | prev:    0.073 | next:    0.079",
  "string:   0.079 | prev:    0.076 | next:    0.082",
  "string:   0.082 | prev:    0.079 | next:    0.086",
  "string:   0.086 | prev:    0.082 | next:     0.09",
  "string:    0.09 | prev:    0.086 | next:    0.094",
  "string:   0.094 | prev:     0.09 | next:    0.098",
  "string:   0.098 | prev:    0.094 | next:    0.102",
  "string:   0.102 | prev:    0.098 | next:    0.106",
  "string:   0.106 | prev:    0.102 | next:     0.11",
  "string:    0.11 | prev:    0.106 | next:    0.114",
  "string:   0.114 | prev:     0.11 | next:    0.118",
  "string:   0.118 | prev:    0.114 | next:    0.124",
  "string:   0.124 | prev:    0.118 | next:     0.13",
  "string:    0.13 | prev:    0.124 | next:    0.136",
  "string:   0.136 | prev:     0.13 | next:    0.142",
  "string:   0.142 | prev:    0.136 | next:    0.148",
  "string:   0.148 | prev:    0.142 | next:    0.154",
  "string:   0.154 | prev:    0.148 | next:     0.16",
  "string:    0.16 | prev:    0.154 | next:    0.168",
  "string:   0.168 | prev:     0.16 | next:    0.176",
  "string:   0.176 | prev:    0.168 | next:    0.184",
  "string:   0.184 | prev:    0.176 | next:    0.192",
  "string:   0.192 | prev:    0.184 | next:      0.2",
  "string:     0.2 | prev:    0.192 | next:    0.208",
  "string:   0.208 | prev:      0.2 | next:    0.216",
  "string:   0.216 | prev:    0.208 | next:    0.226",
  "string:   0.226 | prev:    0.216 | next:    0.236",
  "string:   0.236 | prev:    0.226 | next:    0.246",
  "string:   0.246 | prev:    0.236 | next:    0.256",
  "string:   0.256 | prev:    0.246 | next:    0.266",
  "string:   0.266 | prev:    0.256 | next:     NONE",
]
`;
